### PR TITLE
Graphs Pattern

### PR DESCRIPTION
**Pull request by jbgriesner**
An idiomatic way to implement graphs in Rust.

### CODE REVIEW

File: patterns/graphs.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+Implement a graph data structure that is dynamically modifiable during runtime.
+
+
+## Example
+
+```rust
+// A node of the graph.
+pub struct Node<T> {
+    parent: Option<NodeId>,
```
Comment (FraGag) [inline patterns/graphs.md]: If a node can have at most one parent, then the data structure is a *tree*, not a graph. Nodes in a graph can have more than one parent.

**Comment by nrc** [comment]
This is a way to have a graph in Rust, I'm not sure if its the best way. It's also possible to use `Rc` and `RefCell` with some kind of cycle tracking/detection, or unsafe code in numerous ways.

File: patterns/graphs.md (hunk 1 above)
Comment (lfairy) [reply patterns/graphs.md]: It looks more like a *forest*, since this code allows for multiple roots 😉

**Comment by jbgriesner** [comment]
I made a distinction.
I will code different ways to have graph.

**Comment by lambda-fairy** [comment]
Found this while going through old PRs.

In my view, the intention of this patterns repo is to document situations where there is usually one "best" solution. So an overview of many different graph implementations, while valuable, would not belong here.

I will close this PR. Thanks!

**Comment by simonsan** [comment]
> In my view, the intention of this patterns repo is to document situations where there is usually one "best" solution. So an overview of many different graph implementations, while valuable, would not belong here.
> 
> I will close this PR. Thanks!

Imho a `best` solution is developed over time, I think it could be valuable, as this content doesn't exist in the repository already, to state explicitly that it's not sure to be the best solution but that PRs are welcome to improve the examples or encourage discussion on it to come to a better solution, so I'm reopening it for further inspection.

File: patterns/graphs.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+https://en.wikipedia.org/wiki/Region-based_memory_management
+
+https://github.com/saschagrunert/indextree
+
+https://doc.rust-lang.org/1.1.0/rustc_data_structures/graph/
+
+https://github.com/bluss/petgraph
+
+https://github.com/danigm/rust-graph-example
```
Comment (simonsan) [inline patterns/graphs.md:116]:
```suggestion
- [Region-based memory management](https://en.wikipedia.org/wiki/Region-based_memory_management)
- [Module rustc_data_structures::graph](https://doc.rust-lang.org/1.1.0/rustc_data_structures/graph/)
- [Code example: Arena based tree structure with multithreading support](https://github.com/saschagrunert/indextree)
- [Code example: Graph data structure library](https://github.com/bluss/petgraph)
- [Code example: How to implement complex data structure in rust](https://github.com/danigm/rust-graph-example)
```

File: patterns/graphs.md [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+which is hard to understand and will also lead into runtime borrow checks.
+
+
+## See also
+
+- [Region-based memory management](https://en.wikipedia.org/wiki/Region-based_memory_management)
+- [Module rustc_data_structures::graph](https://doc.rust-lang.org/1.1.0/rustc_data_structures/graph/)
+- [Code example: Arena based tree structure with multithreading support](https://github.com/saschagrunert/indextree)
+- [Code example: Graph data structure library](https://github.com/bluss/petgraph)
+- [Code example: How to implement complex data structure in rust](https://github.com/danigm/rust-graph-example)
```
Comment (simonsan) [inline patterns/graphs.md:115]:
```suggestion
- [Code example: How to implement complex data structure in Rust](https://github.com/danigm/rust-graph-example)
```

**Comment by simonsan** [comment]
But yeah, I know what lambda-fairy wanted to state, it's hard to tell this is a good solution or this not. My guess with the book would probably be a subchapter with graphs and then give some examples 

or

probably better to explain the internal graph data structures as a pattern, something like that:
https://doc.rust-lang.org/1.1.0/src/rustc_data_structures/graph/mod.rs.html#11-403 

and give input on how you would do that on not give some custom implementation. But maybe some recommended crates with already implemented data structures to use, to not reinvent the wheel.

Hmm, thoughts?

**Comment by fade2black** [comment]
I think Graph data structure, on its own, is too wide topic to cover in a couple of pages. As a rule, complexity of algorithms on graphs directly depends their [representation](https://en.wikipedia.org/wiki/Graph_(abstract_data_type)#Representations). What kind of representation we will choose depends on our use case. Factors to take into account: 
- type of graphs: complete graph, trees, directed, undirected, etc.
- compute intensive: speed/complexity of algorithms is important
- scarce memory (embedded systems for example)
- what kind of algorithm we implement: parallel, distributed...
- Is graph dynamic?: we constantly add/remove vertices/edges...
- ... 

Let me give a simple example. Say, I know my graph changes (new edges or vertex) once a day, but the graph is almost complete and in my algorithms I traverse neighbours of a single vertex not too often. I'd probably use adjacency matrix.

**Comment by simonsan** [comment]
I think as well, that it should be a complete chapter of different graph structures, would you agree?

**Comment by fade2black** [comment]
@simonsan I agree. We could categorise into use cases or representations, or anything else what people think. But not everything in a single page. Of course the Rust way should be put on the first place.

**Comment by simonsan** [comment]
> 
> 
> @simonsan I agree. We could categorise into use cases or representations, or anything else what people think. But not everything in a single page. Of course the Rust way should be put on the first place.

I opened an issue for discussions, it's easier going forward I guess

**Comment by pickfire** [comment]
This is too wide and there are few patterns to solve this. But there are also many cases of graph, I think we may need a category rather than writing it like this. For example:

- self-referencing graphs (there are even multiple ways to solve this IIRC and this is a common pitfall I think)
- struct of array vs array of structure
- allocate memory somewhere else and keep reference/index

**Comment by simonsan** [comment]
> 
> 
> This is too wide and there are few patterns to solve this. But there are also many cases of graph, I think we may need a category rather than writing it like this. For example:
> 
>     * self-referencing graphs (there are even multiple ways to solve this IIRC and this is a common pitfall I think)
> 
>     * struct of array vs array of structure
> 
>     * allocate memory somewhere else and keep reference/index

Let's use issue #202 to discuss it, I'll mark this as zombie for the future and close it for now.

**Pull request closed without merging**